facade/logstash.go

```diff
@@ -11,26 +11,73 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// Package agent implements a service that runs on a serviced node. It is
-// responsible for ensuring that a particular node is running the correct services
-// and reporting the state and health of those services back to the master
-// serviced.
-
-package dao
+package facade
 
 import (
-	"github.com/control-center/serviced/domain/servicedefinition"
-	"github.com/control-center/serviced/domain/servicetemplate"
-	"github.com/control-center/serviced/utils"
-
+	"errors"
 	"fmt"
 	"io/ioutil"
 	"os"
 	"path"
 	"runtime"
+	"sync"
 	"strings"
+
+	"github.com/zenoss/glog"
+	"github.com/control-center/serviced/datastore"
+	"github.com/control-center/serviced/domain/servicedefinition"
+	"github.com/control-center/serviced/domain/servicetemplate"
+	"github.com/control-center/serviced/isvcs"
+	"github.com/control-center/serviced/utils"
+	"bytes"
+)
+
+
+var (
+	logstashConfigLock = &sync.Mutex{}
+	ErrLogstashUnchanged = errors.New("logstash config unchanged")
 )
 
+// ReloadLogstashConfig will create a new logstash configuration based on the current
+// templates. If that configuration is different from the one currently used by logstash,
+// then it will restart the logstash container so it can use the new filter set.
+//
+// This method should be called anytime the available service definitions are modified
+// This method depends on the elasticsearch container being up and running.
+func (f *Facade) ReloadLogstashConfig(ctx datastore.Context) error {
+	// serialize updates so that we don't have different threads overwriting the same file.
+	logstashConfigLock.Lock()
+	defer logstashConfigLock.Unlock()
+
+	templates, err := f.GetServiceTemplates(ctx)
+	if err != nil {
+		glog.Errorf("Could not write logstash configuration: %s", err)
+		return err
+	}
+	err = writeLogstashConfiguration(templates)
+	if err == ErrLogstashUnchanged {
+		return nil
+	} else if err != nil {
+		glog.Errorf("Could not write logstash configuration: %s", err)
+		return err
+	} 
+
+	glog.V(2).Infof("Starting logstash container")
```

Comment:
That is old school. As a lot of this PR is copy/paste from older code, I didn't take the time to retrofit the logging.   Moving forward, the plan is to systematically retrofit existing code to fit the standards and remove glog from our tool chain in favor of logrus and logri (the hierarchical logging extension for logrus written by Ian).